cli/command/registry/login.go

```diff
@@ -97,6 +98,16 @@ func verifyloginOptions(dockerCli command.Cli, opts *loginOptions) error {
 		opts.password = strings.TrimSuffix(string(contents), "\n")
 		opts.password = strings.TrimSuffix(opts.password, "\r")
 	}
+
+	if opts.serverAddress != "" {
+		u, err := url.Parse(opts.serverAddress)
+		if err != nil {
+			return errors.Errorf("Invalid server address: %s", opts.serverAddress)
+		}
+		if u.Host == "" {
+			return errors.Errorf("Server address must include a hostname: %s", opts.serverAddress)
+		}
+	}
```

Comment:
Not sure this change is correct; the expected argument is the `hostname[:port]` of the registry (so _only_ a hostname, and optionally a port). I think the fact it accepts a scheme may be an accidental side-effect of how it's handled further on, but this change likely will break "correct" cases, e.g. `docker login docker.io` or `docker login example.com:5000`.

That said; there's definitely ambiguity in some of these areas, so we need to better define these.